scanner/src/main/kotlin/storages/Sw360Storage.kt

```diff
@@ -0,0 +1,169 @@
+                .flatMap { releaseClient.getReleaseById(it.releaseId) }
+                .map { getScanResultOfRelease(it) }
+                .orElse(emptyList())
+                .map { path ->
+                    yamlMapper.readValue<ScanResult>(path.toFile())
+                }
+            Success(ScanResultContainer(id, scanResults))
+        } catch (e: SW360ClientException) {
+            val message = "Could not read scan results for ${id.toCoordinates()} in SW360."
+            log.info { message }
+            Failure(message)
+        }
+
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> {
+        val cachedScanResult = createTempFileForUpload(id)
+
+        return try {
+            yamlMapper.writeValue(cachedScanResult, scanResult)
+
+            val uploadResult = releaseClient.getSparseReleaseByNameAndVersion(createReleaseName(id), id.version)
+                .flatMap { releaseClient.getReleaseById(it.releaseId) }
+                .map { deleteScanResultAttachment(it) }
+                .map { createAttachmentOfScanResult(it, cachedScanResult.toPath()) }
+                .map { releaseClient.uploadAttachments(it) }
+
+            if (uploadResult.isPresent && uploadResult.get().isSuccess) {
+                log.debug { "Stored scan result for '${id.toCoordinates()}' in SW360." }
+                Success(Unit)
+            } else {
+                Failure("Failed to add scan results for '${id.toCoordinates()}' in SW360.")
+            }
+        } catch (e: SW360ClientException) {
+            val message = "Failed to add scan results for '${id.toCoordinates()}' in SW360: " +
+                    e.collectMessagesAsString()
+            log.info { message }
+            Failure(message)
+        } finally {
+            cachedScanResult.safeDeleteRecursively(force = true)
+        }
+    }
+
+    private fun deleteScanResultAttachment(release: SW360Release?): SW360Release =
+        releaseClient.deleteAttachments(
+            release,
+            release?.embedded?.attachments
+                ?.filter { isAttachmentAScanResult(it) }
+                ?.map { it.id }
+        )
+
+    private fun isAttachmentAScanResult(attachment: SW360SparseAttachment): Boolean =
+        attachment.attachmentType == SW360AttachmentType.SCAN_RESULT_REPORT
+                && attachment.filename == SCAN_RESULTS_FILE_NAME
+
+    private fun createReleaseName(id: Identifier): String =
+        listOfNotNull(id.namespace, id.name).joinToString("/")
+
+    private fun getScanResultOfRelease(release: SW360Release): List<Path> {
+        val cachedAttachment = createTempDirectory()
+
+        return try {
+            release.embedded.attachments
+                .filter { isAttachmentAScanResult(it) }
+                .mapNotNull { releaseClient.downloadAttachment(release, it, cachedAttachment).orElse(null) }
+        } finally {
+            cachedAttachment.toFile().safeDeleteRecursively()
```

Comment:
If I understand correctly, the temporary directory with all the files that have just been downloaded is now deleted immediately. So the paths returned by this function point to non-existing files, and deserializing the yml results is going to fail.

I think the logic to delete the temporary files for downloads has to be somehow moved to the readFromStorage() method.

Reply:
You're right, therefore i moved the deletion of the file to `readFromStorage` after creating the `ScanResultContainer`.